pkg/providers/instance/resperrorhandlers_test.go

```diff
@@ -35,15 +39,112 @@ import (
+	errMsgSKUFamilyQuotaFmt            = "subscription level %s vCPU quota for %s has been reached (may try provision an alternative instance type)"
+	errMsgSKUNotAvailableFmt           = "the requested SKU is unavailable for instance type %s in zone %s with capacity type %s, for more details please visit: https://aka.ms/azureskunotavailable"
+	errMsgZonalAllocationFailureFmt    = "unable to allocate resources in the selected zone (%s). (will try a different zone to fulfill your request)"
+	errMsgAllocationFailureFmt         = "unable to allocate resources with selected VM size (%s). (will try a different VM size to fulfill your request)"
+	errMsgOverconstrainedZonalFmt      = "unable to allocate resources in the selected zone (%s) with %s capacity type and %s VM size. (will try a different zone, capacity type or VM size to fulfill your request)"
+	errMsgOverconstrainedAllocationFmt = "unable to allocate resources in all zones with %s capacity type and %s VM size. (will try a different capacity type or VM size to fulfill your request)"
+	errMsgRegionalQuotaExceeded        = "regional on-demand vCPU quota limit for subscription has been reached. To scale beyond this limit, please review the quota increase process here: https://learn.microsoft.com/en-us/azure/quotas/regional-quota-requests"
+)
+
+// offering represents a zone and capacity type combination for cleaner test setup
+type offering struct {
+	zone         string
+	capacityType string
+}
+
+// Common offering configurations
+var (
+	zone1OnDemand = offering{zone: testZone1, capacityType: karpv1.CapacityTypeOnDemand}
+	zone1Spot     = offering{zone: testZone1, capacityType: karpv1.CapacityTypeSpot}
+	zone2OnDemand = offering{zone: testZone2, capacityType: karpv1.CapacityTypeOnDemand}
+	zone2Spot     = offering{zone: testZone2, capacityType: karpv1.CapacityTypeSpot}
+	zone3OnDemand = offering{zone: testZone3, capacityType: karpv1.CapacityTypeOnDemand}
+	zone3Spot     = offering{zone: testZone3, capacityType: karpv1.CapacityTypeSpot}
 )
 
+type testCaseBuilder struct {
+	tc responseErrorTestCase
+}
+
+func newTestCase(name string) *testCaseBuilder {
+	return &testCaseBuilder{
+		tc: responseErrorTestCase{
+			testName:                                name,
+			originalRequestSKU:                      createDefaultTestSKU(),
+			expectedUnavailableOfferingsInformation: []offeringToCheck{},
+			expectedAvailableOfferingsInformation:   []offeringToCheck{},
+		},
+	}
+}
+
+func (b *testCaseBuilder) withInstanceType(offerings ...offering) *testCaseBuilder {
+	offeringTypes := make([]struct{ zone, capacityType string }, len(offerings))
+	for i, o := range offerings {
+		offeringTypes[i] = struct{ zone, capacityType string }{zone: o.zone, capacityType: o.capacityType}
+	}
+	b.tc.instanceType = defaultCreateInstanceType(offeringTypes...)
```

Comment:
Should this helper `defaultCreateInstanceType` just take `offering`? Bascially go through your whole test here and drop `struct{ zone, capacityType string }` in favor of the named type?

Reply:
Good point, this test file went through like 10 iterations. Changing to use `offering`